Embedded Typescript GitHub Action

josecelano opened the pull request:
I've created the basic scaffolding to extract the new GitHub Action using node 16.

99% of the code is a copy/paste from https://github.com/crazy-max/ghaction-import-gpg

I'm interested in defining the action API.

josecelano commented:
The interesting things are here:

- [Action definition](https://github.com/josecelano/library-consumer/blob/issue-7-extract-gh-actions/.github/actions/git-queue/action.yml)
- [Sample workflow using the action](https://github.com/josecelano/library-consumer/blob/issue-7-extract-gh-actions/.github/workflows/test-git-queue-action.yml)

da2ce7 commented:
Is there a standard way to separate the compiled `.github/actions/git-queue/dist/index.js` from the pull request?

josecelano commented:
> Is there a standard way to separate the compiled `.github/actions/git-queue/dist/index.js` from the pull request?

It seems to be the recommended way to do it. They [recommend](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action#commit-tag-and-push-your-action-to-github) to use [@vercel/ncc](https://github.com/vercel/ncc). It's only one file so you can ignore it.

I suppose we could ignore the dist folder in a `development` branch and then build the final action when we merge into `main`. We can do that once we move the action to its own repo.

Even [Actions](https://github.com/actions/setup-node) from the GitHub team have the package app inside the repo.

Other repos from GitHub: https://github.com/orgs/actions/repositories?type=all

By the way, I found the [official GitHub Action template for TypeScript](https://github.com/actions/typescript-action). It looks pretty similar to what I'm using, but I will check it.

da2ce7 commented:
> By the way, I found the [official GitHub Action template for TypeScript](https://github.com/actions/typescript-action). It looks pretty similar to what I'm using, but I will check it.

I think that this template looks quite good. :)

josecelano commented:
hi, @da2ce7 I've implemented one of the cases for the action:

- [ ] Create job
- [x] Get next job (get the current pending job)
- [ ] Mark job as done

I've created a [Queue](https://github.com/josecelano/library-consumer/blob/issue-7-extract-gh-actions/.github/actions/git-queue/src/queue.ts) class.

The queue does not care about the payload, so we still need to parse the message payload to get the variables we need for the job (`current_ref`, `previous_ref`). The payload is the commit message body.

I will continue with the other two cases (create git commits for new-job commit and for the job-done commit).

If we use a json payload I think I could parse the variables using `jq`. FOr example the `worker` workflow could be like this:

```
...
- name: Get next job
  id: get-next-job
  uses: ./.github/actions/git-queue
  with:
    queue_name: "Library Update [library-aaa]"
    action: "next-job"

- name: Get next job
  id: parse-job-payload
  run: |
    $PREVIOUS_REF = echo ${{ steps.get-next-job.job_payload }} | jq '.previous_ref'
    $CURRENT_REF = echo ${{ steps.get-next-job.job_payload }} | jq '.current_ref'
    echo "::set-output name=previous_ref::$PREVIOUS_REF"
    echo "::set-output name=current_ref::$CURRENT_REF"
    echo "::set-output name=active_lock::true"
...
```

We have to generate this payload:

```json
{
  "name": "",
  "version": "1",
  "data": {
    "previous_ref": "asfa8979sd7f",
    "current_ref": "asfa8979sd7f"
  } 
}
```

The payload is defined by the queue user. We probably should define another json for the commit body in a way that we can also store metadata in the future, not only the user payload.

```json
{
  "version": "1",
  "metadata": {
    ...
  },
  "payload": "escaped string with the user payload"
}
```

I can also finish this first implementation using a json payload but only storing the user payload in the commit body, for the time being.

josecelano commented:
@da2ce7 I've added another case: create a new job. I'm going to continue with "Mark job as done".

josecelano commented:
I've also added the last case "mark job as done". I would like to add more tests but I want to run first some integrations tests, replacing the new action in the workflow.

josecelano commented:
hi @da2ce7 @yeraydavidrodriguez ,

The [test workflow](https://github.com/josecelano/library-consumer/blob/issue-7-extract-gh-actions/.github/workflows/test-git-queue-action.yml) is working.

You can see [here](https://github.com/josecelano/library-consumer/runs/4956813133?check_suite_focus=true#step:12:1) the last step with the new commits.

TODO:

- [ ] [Define the commit subject and body](https://github.com/josecelano/library-consumer/issues/4). And make the necessary changes.
- [ ] I've added a test to test signed commits, but it only works locally if you have imported the GPG key like in the test workflow. I can add the setup to import the GPG key as we did for the [librarian](https://github.com/Nautilus-Cyberneering/nautilus-librarian/blob/main/tests/test_nautilus_librarian/test_mods/test_gpg/fixtures/gpg_fixtures.py#L86).
- [ ] Add more tests once we define the final behavior.

QUESTIONS:

- Should we always "push" after the commit? Maybe we can use an input variable in order to let the users decide whether they want to push immediately or not.
- Should I implement the `MarkJobAsStarted` message.
- Should we make multi-jobs?

josecelano commented:
> 
> TODO:
> 
> * [x]  [Define the commit subject and body](https://github.com/josecelano/library-consumer/issues/4). And make the necessary changes.
> * [x]  I've added a test to test signed commits, but it only works locally if you have imported the GPG key like in the test workflow. I can add the setup to import the GPG key as we did for the [librarian](https://github.com/Nautilus-Cyberneering/nautilus-librarian/blob/main/tests/test_nautilus_librarian/test_mods/test_gpg/fixtures/gpg_fixtures.py#L86).
> * [x]  Add more tests once we define the final behavior.
> * [x]  Add a new input variable with the signing key. The test workflow is not using signed commits yet.
> 
> QUESTIONS:
> 
> * Should we always "push" after the commit? Maybe we can use an input variable in order to let the users decide whether they want to push immediately or not.
> * Should we implement the `MarkJobAsStarted` message? Yes
> * Should we implement multi-jobs queue? Yes
> 
> NEXT STEPS (after TODO):
> 
> * Move the repo to the organization using [GitHub official TypeScript template](https://github.com/actions/typescript-action)
> * Use the action in the [website repo](https://github.com/Nautilus-Cyberneering/chinese-ideographs-website).
> * Improve actions docs

hi @da2ce7 @yeraydavidrodriguez I think this PR is ready to merge. The latest changes are:

- Setup for importing GPG key for testing.
- Added two arguments to the action that match `git commit` arguments in order to allow to customize those arguments.
- Do not run `git push` inside the action.
- Added new tests.
- Allow passing env vars to the git child process.

The hardest part was to preset the passphrase for the GPG key (again :-)) because it's very tricky to synchronize the GPG Agent with the test and I had to change the code I copied from the "GPG Key Import Action" too because it did not allow changing the GPG `homedir`.

### Nest steps

1. Use the embedded action in the `worker` and `work-allocator` workflows in this repo.
2. Extract the action into its own repo

In the new repo:

1. New GitHub Action template
2. Add mega-linter
3. New issue: new conventions definition from https://github.com/josecelano/library-consumer/issues/4 (only definition, docs).
3. New issue: refactor current messages to use new wording (only wording, no protocol literals).
4. New issue: add new message "Finish Job".
5. New issue: refactor to use json payload.
6. New issue: change protocol keywords (commit message)

On the website repo:

1. Use the action

da2ce7 commented:
Hello Jose,

Sounds good. Let’s go ahead with the merge.

Are there any negative security implications for having the pgp-sub-key without any passphase? Maybe this would make things more simple?

:)

Cam.

Von meinem iPhone gesendet

> Am 1/2/2022 um 14:56 schrieb Jose Celano ***@***.***>:
> 
> ﻿
> TODO:
> 
>  Define the commit subject and body. And make the necessary changes.
>  I've added a test to test signed commits, but it only works locally if you have imported the GPG key like in the test workflow. I can add the setup to import the GPG key as we did for the librarian.
>  Add more tests once we define the final behavior.
>  Add a new input variable with the signing key. The test workflow is not using signed commits yet.
> QUESTIONS:
> 
> Should we always "push" after the commit? Maybe we can use an input variable in order to let the users decide whether they want to push immediately or not.
> Should we implement the MarkJobAsStarted message? Yes
> Should we implement multi-jobs queue? Yes
> NEXT STEPS (after TODO):
> 
> Move the repo to the organization using GitHub official TypeScript template
> Use the action in the website repo.
> Improve actions docs
> hi @da2ce7 @yeraydavidrodriguez I think this PR is ready to merge. The latest changes are:
> 
> Setup for importing GPG key for testing.
> Added two arguments to the action that match git commit arguments in order to allow to customize those arguments.
> Do not run git push inside the action.
> Added new tests.
> Allow passing env vars to the git child process.
> The hardest part was to preset the passphrase for the GPG key (again :-)) because it's very tricky to synchronize the GPG Agent with the test and I had to change the code I copied from the "GPG Key Import Action" too because it did not allow changing the GPG homedir.
> 
> Nest steps
> 
> Use the embedded action in the worker and work-allocator workflows in this repo.
> Extract the action into its own repo
> In the new repo:
> 
> New GitHub Action template
> Add mega-linter
> New issue: new conventions definition from #4 (only definition, docs).
> New issue: refactor current messages to use new wording (only wording, no protocol literals).
> New issue: add new message "Finish Job".
> New issue: refactor to use json payload.
> New issue: change protocol keywords (commit message)
> On the website repo:
> 
> Use the action
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you were mentioned.

yeraydavidrodriguez commented on .github/actions/git-queue/__tests__/email-address.test.ts:
```diff
@@ -0,0 +1,30 @@
+    const t = () => {
+      new EmailAddress('abc_NO_AT_example.com');
+    };
+    expect(t).toThrow(Error);
```
If the expectation is to throw an error, shouldn't the name of the test "should NOT validate an invalid email address?

yeraydavidrodriguez reviewed (approved):
Good job, just a minor comment

josecelano commented:
> Hello Jose, Sounds good. Let’s go ahead with the merge. Are there any negative security implications for having the pgp-sub-key without any passphase? Maybe this would make things more simple? :) Cam.

hi @da2ce7 

1. The action itself does not deal with GPG keys. The action user should make sure that the GPG keys are set up correctly before using the action. I'm using GPG keys (import and sign) only the tests. In fact, I only use signed commits in 2 tests. I was using the passphrase because it's the way we are importing GPG keys in other projects and I wanted to use a "real" or "common" scenario for the test. If we have concurrency problems in the future we can remove it.
2. Regarding security. To be honest, I'm not sure if using the passphrase makes sense. If I'm not wrong, it should add an extra security barrier because we copy the encrypted private key into the runner but the passphrase is injected in the runner only in memory (main shell user process). But I have not confirmed that yet. I did a fast search and I did not find documentation about the GitHub internal workflow cycle.
3. Regarding how to make things simpler I also have some doubts about that. We delegate the "git push" to the action user but I'm wondering if we shouldn't even make the commit. I mean the action could only build the commit message and we could let the user commit. That way we do not have to deal with passing some arguments from the user to the git command. The users can setup git and commit whatever they like. I have the feeling that we can end up having all the "git commit" arguments as action inputs. But anyway I think we cover 99% of the common use cases. We could also return the message and add a boolean input to disable the actual commit. If we do not deal with "git commit" the test scaffolding would be definitively simpler.
